### ISSUE TITLE
Query syntax issue

### ISSUE THREAD

**Issue by davidagold** [davidagold/StructuredQueries.jl#21]
This is a continuation of #16 , which this issue expands on with other syntax issues/desiderata.

There we considered how piping should be expressed: by use of newlines (within a `begin`-`end` block) or by use of `|>`. I currently support the case of "both". Users should be able to express a pipe in-line by `|>`. But oftentimes, one does `|>` and a newline, which is redundant. Users should be able to let newlines express piping if they choose. I find the latter scheme declutters the query syntax.

If we allow newlines to express piping, then the following is allowable syntax:

``` julia
qry = @query begin
    tbl1
        select(A)
end
```

That is, a line containing just a symbol (e.g. `tbl1`) marks the beginning of a flow of data (from `tbl1`) through a series of manipulation verbs (e.g. `select(A)`), each of which refers to the data source named by the symbol that began the flow. The flow is ended by the end of the block. 

We also want to support:
1. Queries against multiple backends 
2. Queries that refer to [CTEs](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression), or something like them 

Which suggests allowing for multiple such manipulation flows within an `@query` invocation as well as providing a means of referring to the columns of another table. A natural option for the latter is the `alias.column` format of identifier. For instance, in the following, `_tbl` is an alias for the result of the "subquery" consisting in the manipulations that are begun by `tbl1`. The `filter` command in the manipulations pertaining to `tbl2` references the `B` column of `_tbl` by means of `_tbl.B`. 

``` julia
qry = @query begin
    _tbl = tbl1
        filter(A > .5)
        select(B)
    tbl2
        filter(B > _tbl.B)
        select(A)
end
```

**Comment by yeesian** [comment]
Would

(A)

``` julia
qry = @query begin
    tbl1 select(A)
end
```

(B)

``` julia
qry = @query begin
    tbl1 filter(A > .5) select(B)
end
```

(C)

``` julia
qry = @query begin
    tbl1 filter(A > .5)
        select(B)
end
```

be considered valid syntax?

**Comment by davidagold** [comment]
They can't be, as they're not valid Julia syntax. I do like the look, though.

**Comment by nalimilan** [comment]
I'd prefer each subquery to be a `source begin... end` block. There's no other place in Julia where a variable name can serve as a new block marker.

**Comment by davidagold** [comment]
@nalimilan True, but then, this isn't regular Julia code; `select`, etc. are not functions that are ever called. So I'm not opposed in principle to departing from normal Julia syntax, since this is a custom DSL.

**Comment by yeesian** [comment]
What about two different kinds of begin blocks:
- the first

``` julia
qry = @query src |> begin
    filter(...)
    select(...)
end
```

as a stand-in substitute for piping

``` julia
qry = @query src |>
    filter(...) |>
    select(...)
```

(**Remark**: observe the pipe `|>` before the `begin`)
- the second for more complex queries that might involve the construction of CTEs, etc:

(using pipes)

``` julia
qry = @query begin
    _tbl = tbl1 |>
        filter(A > .5) |>
        select(B)
    tbl2 |>
        filter(B > _tbl.B) |>
        select(A)
end
```

or equivalently (using blocks)

``` julia
qry = @query begin
    _tbl = tbl1 |> begin
        filter(A > .5)
        select(B)
    end
    tbl2 |> begin
        filter(B > _tbl.B)
        select(A)
    end
end
```

or equivalently (using a mix)

``` julia
qry = @query begin
    _tbl = tbl1 |>
        filter(A > .5) |>
        select(B)
    tbl2 |> begin
        filter(B > _tbl.B)
        select(A)
    end
end
```

**Comment by nalimilan** [comment]
@davidagold Yeah, that's not an absolute requirement, but better remain as close as possible to standard Julia code if it suits our needs.

**Comment by davidagold** [comment]
@nalimilan It's a little difficult, though, since

``` julia
@query tbl1 begin
    filter(A > .5)
end
    tbl2 begin
    filter(B > .5)
end
```

won't be properly visible to the macro -- one would need to do

``` julia
@query tbl1 begin
    filter(A > .5)
end tbl2 begin
    filter(B > .5)
end
```

and I'm not a huge fan of (i) the proliferation of `begin`/`end` invocations necessary to describe manipulations of multiple sources, (ii) having a source name be stuck between two such blocks on the same line. 

I think I prefer that `@query begin ... end` introduces a "query context" in which a special domain-specific syntax applies. This syntax will definitely include special conventions for identifying column and table names, and it may include special conventions for beginning code blocks. But users should always keep in mind that it is not a "plain" Julia context.
